packages/flutter/lib/src/widgets/pages.dart

```diff
@@ -128,3 +128,85 @@ class PageRouteBuilder<T> extends PageRoute<T> {
   }
 
 }
+
+/// A [Navigator] observer that notifies [PageRouteAware]s of changes to the
+/// state of their [PageRoute].
+///
+/// To make a [StatefulWidget] aware of its current [PageRoute] state, implement
+/// [PageRouteAware] in its [State] and subscribe it to the [PageRouteObserver]:
+///
+/// ```dart
+/// class PageRouteAwareWidgetState extends State<PageRouteAwareWidget> with PageRouteAware {
+///   @override
+///   void didChangeDependencies() {
+///     super.didChangeDependencies();
+///     pageRouteObserver.subscribe(this, ModalRoute.of(context));
+///   }
+///
+///   @override
+///   void dispose() {
+///     pageRouteObserver.unsubscribe(this);
+///     super.dispose();
+///   }
+///
+///   @override
+///   void didPush() {
+///     // Do something
+///   }
+///
+///   @override
+///   void didPopNext() {
+///     // Do something
+///   }
+///
+/// }
+/// ```
+class PageRouteObserver extends NavigatorObserver {
```

Comment:
This could theoretically be a templated class `RouteObserver<T extends Route>` if you wanted to track other transitions besides PageRoutes, I guess. What do you think?

LGTM otherwise